### DISCUSSION TITLE
Can PMD identify circular references across packages?

### DISCUSSION THREAD

**Discussion by vrmerlin** [pmd/pmd#4406] [Q&A]
We need to break up a very large code base into multiple repos. Does PMD provide a way to identify circular references between packages/classes, so we can address those problems in the code?

**Comment by jsotuyod** [comment]
Hi @vrmerlin. There is currently no rule to detect this, but it would be possible to write such a rule (I actually think someone already tried something like this at some point). Albeit, currently writing such a rule would not be fully compliant (ie: incremental analysis would not work with it), but we have plans to eventually resolve the underlaying issues preventing this…

Right now it would require a [custom Java Rule](https://pmd.sourceforge.io/pmd-6.54.0/pmd_userdocs_extending_writing_java_rules.html), and having a `static` shared state (this is the non-compliant piece) where to store for each class everything they are referencing, and looking for cycles (be careful for concurrency issues on this static shared state) to report.

**Reply by NilsBaumgartner1994** [reply]
See https://github.com/pmd/pmd/issues/3920 and https://github.com/pmd/pmd/discussions/4565 as it requests such support.
